File: android/app/src/main/java/br/com/app/gpu2675756/gpu0e7509bfb7bde52aef412888bb17a456/NativeVadeMecumPlugin.kt
package br.com.app.gpu2675756.gpu0e7509bfb7bde52aef412888bb17a456

import android.content.Intent
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "NativeVadeMecumPlugin")
class NativeVadeMecumPlugin : Plugin() {

    companion object {
        var activeCall: PluginCall? = null
        var instance: NativeVadeMecumPlugin? = null
    }

    override fun load() {
        super.load()
        instance = this
    }

    @PluginMethod
    fun openArtigo(call: PluginCall) {
        val id = call.getString("id") ?: ""
        val numero = call.getString("numero") ?: ""
        val caput = call.getString("caput") ?: ""
        val titulo = call.getString("titulo") ?: ""
        val tabelaNome = call.getString("tabelaNome") ?: ""
        val paragrafos = call.getArray("paragrafos")?.toString() ?: "[]"
        val incisos = call.getArray("incisos")?.toString() ?: "[]"
        val highlights = call.getArray("highlights")?.toString() ?: "[]"
        val audioUrl = call.getString("audioUrl") ?: ""
        val accessToken = call.getString("accessToken") ?: ""

        activeCall = call

        val intent = Intent(context, ArtigoNativeActivity::class.java).apply {
            putExtra("id", id)
            putExtra("numero", numero)
            putExtra("caput", caput)
            putExtra("titulo", titulo)
            putExtra("tabelaNome", tabelaNome)
            putExtra("paragrafos", paragrafos)
            putExtra("incisos", incisos)
            putExtra("highlights", highlights)
            putExtra("audioUrl", audioUrl)
            putExtra("accessToken", accessToken)
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }

        context.startActivity(intent)
        call.resolve()
    }

    fun emitHighlightsUpdated(artigoId: String, highlightsJson: String) {
        val ret = JSObject().apply {
            put("artigoId", artigoId)
            put("highlights", highlightsJson)
        }
        notifyListeners("onHighlightsUpdated", ret)
    }
}

File: ios/App/App/NativeVadeMecumPlugin.swift
import Capacitor
import SwiftUI

@objc(NativeVadeMecumPlugin)
public class NativeVadeMecumPlugin: CAPPlugin {
    
    public override func load() {
        super.load()
    }

    @objc func openArtigo(_ call: CAPPluginCall) {
        let id = call.getString("id") ?? ""
        let numero = call.getString("numero") ?? ""
        let caput = call.getString("caput") ?? ""
        let titulo = call.getString("titulo") ?? ""
        let tabelaNome = call.getString("tabelaNome") ?? ""
        let paragrafos = call.getArray("paragrafos", String.self) ?? []
        let incisos = call.getArray("incisos", String.self) ?? []
        let audioUrl = call.getString("audioUrl") ?? ""
        
        let initialHighlights = (call.getArray("highlights", JSObject.self) ?? []).compactMap { obj -> String? in
            guard let line = obj["lineIndex"] as? Int,
                  let word = obj["wordIndex"] as? Int,
                  let color = obj["colorHex"] as? String else { return nil }
            return "\(line):\(word):\(color)"
        }
        
        let artigoData = ArtigoData(
            id: id,
            numero: numero,
            caput: caput,
            titulo: titulo,
            tabelaNome: tabelaNome,
            paragrafos: paragrafos,
            incisos: incisos,
            audioUrl: audioUrl,
            initialHighlights: initialHighlights
        )
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let rootVC = self.bridge?.viewController else {
                call.reject("Unable to find root view controller")
                return
            }
            
            var hostingController: UIHostingController<ArtigoView>?
            
            let artigoView = ArtigoView(
                artigo: artigoData,
                onClose: {
                    hostingController?.dismiss(animated: true, completion: nil)
                },
                onHighlightsChanged: { [weak self] updatedHighlights in
                    self?.notifyListeners("onHighlightsUpdated", data: [
                        "artigoId": id,
                        "highlights": updatedHighlights
                    ])
                }
            )
            
            hostingController = UIHostingController(rootView: artigoView)
            hostingController?.modalPresentationStyle = .fullScreen
            
            if let hc = hostingController {
                rootVC.present(hc, animated: true) {
                    call.resolve()
                }
            }
        }
    }
}
